hal/src/common/thumbv6m/sercom/v2/spi.rs

```diff
@@ -1428,6 +1438,28 @@ impl<C: ValidConfig> Spi<C> {
         &self.config.as_ref().sercom()
     }
 
+    /// Update the SPI configuration.
+    ///
+    /// Calling this method will temporarily disable the SERCOM peripheral, as
+    /// some registers are enable-protected. This may interrupt any ongoing
+    /// transactions.
+    #[inline]
+    pub fn update_config<F>(&mut self, update: F)
+    where
+        F: FnOnce(C) -> C,
```

Comment:
I think I would prefer `F: FnOnce(SpecificConfig<C>) -> SpecificConfig<C>,`. It probably doesn't matter that much, because you're probably only calling this function when you have a concrete type for `C`, i.e. in end-user code. But `SpecificConfig<C>` tells the compiler that it will definitely receive a `Config` struct, not some unknown, arbitrary type `C`.

Reply:
That doesn't seem to compile, because the `Spi` struct expects `self.config` to be of type `C`, and so `replace_with` fails to compile when trying to replace some type `C` with a specific `Config`.

Reply:
@jbeaurivage, yes this is a bit tricky. I mention a similar situation in the [docs](https://atsamd-rs.github.io/docs/samd51n/thumbv7em-none-eabihf/doc/atsamd_hal/common/gpio/v2/pin/trait.AnyPin.html#generic-anypins) for `AnyPin`.

The compiler doesn't understand that when `C: AnyConfig`, `C == SpecificConfig<C>>`. From the compiler's perspective, `C` is some unknown, generic type, while `SpecificConfig<C>` is a `Config` built from associated types. However, `C` implements `AnyConfig`, which implies the bounds `C: From<SpecificConfig<C>>` and `C: Into<SpecificConfig<C>>`, so you can trivially "convert" between them, which is always a NOP.

The solution is

```rust
replace_with::replace_with(
    &mut self.config,
    || old_config,
    |config| update(config.into()).into()
);
```

I just checked, and this compiles.